### DISCUSSION TITLE
支持动态upstream

### DISCUSSION THREAD

**Discussion by marchlhw** [bfenetworks/bfe#1171] [General]
后端实例ip+port不需要人为配置，直接从注册中心根据相关前缀key获取，可以支持常用的注册中心，比如：etcd、zk、eureka、nacos等

**Comment by githublaohu** [comment]
etcd、zk、eureka、nacos。全部要支持吗？

**Comment by arrow2012** [comment]
动态upstream 主流的 用 consul，ETCD  比较多，这2个应该要支持下

**Comment by arrow2012** [comment]
动态upstream 这个功能预计什么时候出来？ @mileszhang2016

**Comment by mileszhang2016** [comment]
@arrow2012 
目前一个百度之外的同学在开发支持nacos的组件
初稿在这里： https://github.com/githublaohu/bfe-register
等打磨好后，会放入bfenetworks中。
也非常欢迎更多社区的同学来贡献支持 etcd、zk、eureka 的组件。

**Comment by liuqing6767** [comment]
这个功能实现思路大体上有：
1. 数据面直接和注册中心对接，这个和现有配置加载逻辑相差太大，暂不考虑
1. 控制面 和 注册中心对接，保持控制面加载逻辑
    1. 单独提供 probe 来做 注册中心和APISserver之间的数据同步
    1. 在 APIServer中通过对 [PoolStorage](https://github.com/bfenetworks/api-server/blob/develop/model/icluster_conf/pool.go#L76) 进行多态实现

考虑到运维成本、实现成本，建议使用   2.2 方案

**Comment by liuqing6767** [comment]
这个issue 的 api-server 版本实现进行中，https://github.com/bfenetworks/api-server/projects/4
